From 6b2c34bbb2541abdd67eaf0d713ffcf162c479d5 Mon Sep 17 00:00:00 2001
From: Alejandro del Castillo <alejandro.delcastillo@ni.com>
Date: Mon, 13 Jan 2020 17:19:55 -0600
Subject: [PATCH 1/2] opkg_install_pkg: allow upgrades between packages with
 different names

Required to be able to process OBSOLETE operations as upgrades.

Signed-off-by: Alejandro del Castillo <alejandro.delcastillo@ni.com>

Upstream-status: Submitted
---
 libopkg/opkg.c                                   | 2 +-
 libopkg/opkg_install.c                           | 6 +++---
 libopkg/opkg_install.h                           | 2 +-
 libopkg/solvers/internal/opkg_install_internal.c | 4 ++--
 libopkg/solvers/libsolv/opkg_solver_libsolv.c    | 4 ++--
 5 files changed, 9 insertions(+), 9 deletions(-)

diff --git a/libopkg/opkg.c b/libopkg/opkg.c
index 05de83db63c76f4b5d1d6a79f67f4b2164f82b58..037f4707954c162237ed23f1da6f2ac1f481a03b 100644
--- a/libopkg/opkg.c
+++ b/libopkg/opkg.c
@@ -311,7 +311,7 @@ int opkg_install_package(const char *package_url,
     progress(&pdata, 75, progress_callback, user_data);
 
     /* unpack the package */
-    err = opkg_install_pkg(new);
+    err = opkg_install_pkg(new, NULL);
 
     if (err) {
         return -1;
diff --git a/libopkg/opkg_install.c b/libopkg/opkg_install.c
index c1e7e048416f8f2e0e1d88634ee2ed645012a438..e7a9a24adcbb2ca93542d09a8ad2d3e36f8d51f0 100644
--- a/libopkg/opkg_install.c
+++ b/libopkg/opkg_install.c
@@ -881,10 +881,9 @@ static int resolve_conffiles(pkg_t * pkg)
 /**
  *  @brief Really install a pkg_t
  */
-int opkg_install_pkg(pkg_t * pkg)
+int opkg_install_pkg(pkg_t * pkg, pkg_t * old_pkg)
 {
     int err = 0;
-    pkg_t *old_pkg = NULL;
     abstract_pkg_t *ab_pkg = NULL;
     int old_state_flag;
     sigset_t newset, oldset;
@@ -902,7 +901,8 @@ int opkg_install_pkg(pkg_t * pkg)
         pkg->dest = opkg_config->default_dest;
     }
 
-    old_pkg = pkg_hash_fetch_installed_by_name(pkg->name);
+    if (!old_pkg)
+        old_pkg = pkg_hash_fetch_installed_by_name(pkg->name);
 
     pkg->state_want = SW_INSTALL;
     if (old_pkg) {
diff --git a/libopkg/opkg_install.h b/libopkg/opkg_install.h
index 98378bb02b37c82e00d0bc5142994d9c627cf70a..06246d93cfd24b0bbe8710f79b8f862282615b5c 100644
--- a/libopkg/opkg_install.h
+++ b/libopkg/opkg_install.h
@@ -28,7 +28,7 @@
 extern "C" {
 #endif
 
-int opkg_install_pkg(pkg_t * pkg);
+int opkg_install_pkg(pkg_t * pkg, pkg_t * old_pkg);
 
 #ifdef __cplusplus
 }
diff --git a/libopkg/solvers/internal/opkg_install_internal.c b/libopkg/solvers/internal/opkg_install_internal.c
index 59cbe0956636f5407a03a481bf044d98da2a780e..a33fa518d3fd8f98eb90e753c53c0cb5c1dbd74f 100644
--- a/libopkg/solvers/internal/opkg_install_internal.c
+++ b/libopkg/solvers/internal/opkg_install_internal.c
@@ -54,7 +54,7 @@ static int pkg_remove_installed_replacees_unwind(pkg_vec_t *replacees)
         pkg_t *replacee = replacees->pkgs[i];
         if (replacee->state_status != SS_INSTALLED) {
             opkg_msg(DEBUG2, "Calling opkg_install_pkg.\n");
-            err = opkg_install_pkg(replacee);
+            err = opkg_install_pkg(replacee, NULL);
             if (err)
                 return err;
         }
@@ -193,7 +193,7 @@ int opkg_execute_install(pkg_t *pkg, pkg_vec_t *pkgs_to_install, pkg_vec_t *repl
         /* Set all pkgs to auto_installed except the top level */
         if (dependency != pkg)
             dependency->auto_installed = 1;
-        r = opkg_install_pkg(dependency);
+        r = opkg_install_pkg(dependency, NULL);
         if (r < 0)
             errors++;
     }
diff --git a/libopkg/solvers/libsolv/opkg_solver_libsolv.c b/libopkg/solvers/libsolv/opkg_solver_libsolv.c
index 373abe8334ab861a906535c17739437e2828036d..d00611132c4eeac08351d04dfb784bab67564088 100644
--- a/libopkg/solvers/libsolv/opkg_solver_libsolv.c
+++ b/libopkg/solvers/libsolv/opkg_solver_libsolv.c
@@ -953,7 +953,7 @@ static int libsolv_solver_execute_transaction(libsolv_solver_t *libsolv_solver)
                     opkg_message(NOTICE, "Installing %s (%s) on %s\n",
                                  pkg->name, pkg->version, pkg->dest->name);
                 }
-                ret = opkg_install_pkg(pkg);
+                ret = opkg_install_pkg(pkg, NULL);
                 if (ret) {
                     err = -1;
                     goto CLEANUP;
@@ -985,7 +985,7 @@ static int libsolv_solver_execute_transaction(libsolv_solver_t *libsolv_solver)
                     }
                 }
 
-                ret = opkg_install_pkg(pkg);
+                ret = opkg_install_pkg(pkg, NULL);
                 if (ret) {
                     err = -1;
                     goto CLEANUP;
-- 
2.25.0.rc2

